lib/wasi/src/runtime/mod.rs

```diff
@@ -0,0 +1,347 @@
+mod stdio;
+pub mod task_manager;
+mod ws;
+
+use self::task_manager::StubTaskManager;
+pub use self::{
+    stdio::*,
+    task_manager::{SpawnType, SpawnedMemory, VirtualTaskManager},
+    ws::*,
+};
+
+use std::{
+    fmt,
+    future::Future,
+    io::{self, Write},
+    pin::Pin,
+    sync::Arc,
+};
+
+use thiserror::Error;
+use tracing::*;
+use wasmer_vbus::{DefaultVirtualBus, VirtualBus};
+use wasmer_vnet::{DynVirtualNetworking, VirtualNetworking};
+use wasmer_wasi_types::wasi::Errno;
+
+use crate::{os::tty::WasiTtyState, WasiEnv};
+
+#[cfg(feature = "termios")]
+pub mod term;
+use crate::http::DynHttpClient;
+#[cfg(feature = "termios")]
+pub use term::*;
+
+#[cfg(feature = "sys")]
+pub type ArcTunables = std::sync::Arc<dyn wasmer::Tunables + Send + Sync>;
+
+#[derive(Error, Debug)]
+pub enum WasiThreadError {
+    #[error("Multithreading is not supported")]
+    Unsupported,
+    #[error("The method named is not an exported function")]
+    MethodNotFound,
+    #[error("Failed to create the requested memory")]
+    MemoryCreateFailed,
+    /// This will happen if WASM is running in a thread has not been created by the spawn_wasm call
+    #[error("WASM context is invalid")]
+    InvalidWasmContext,
+}
+
+impl From<WasiThreadError> for Errno {
+    fn from(a: WasiThreadError) -> Errno {
+        match a {
+            WasiThreadError::Unsupported => Errno::Notsup,
+            WasiThreadError::MethodNotFound => Errno::Inval,
+            WasiThreadError::MemoryCreateFailed => Errno::Fault,
+            WasiThreadError::InvalidWasmContext => Errno::Noexec,
+        }
+    }
+}
```

Comment:
Can we move this where WasiThreads are defined? (I think it was near the control plane?)

Reply:
Moved to `os::task::thread` submodule in a155340b1caf50935f2741e614b35e68f5ad4b02 .